### PR TITLE
Using MarkGood and StopPeerForError

### PR DESCRIPTION
**Pull request by melekes**
Refs #1147

### CODE REVIEW

File: blockchain/reactor.go [hunk 1]
```diff
@@ -22,8 +22,7 @@ const (
 	// BlockchainChannel is a channel for blocks and status updates (`BlockStore` height)
 	BlockchainChannel = byte(0x40)
 
-	defaultChannelCapacity = 1000
```
Comment (ebuchman) [inline blockchain/reactor.go]: why not leave this here if we still using it ? bad naming ?

File: consensus/reactor.go [hunk 2]
```diff
@@ -1055,6 +1111,36 @@ func (ps *PeerState) ensureVoteBitArrays(height int64, numValidators int) {
 	}
 }
 
+// RecordVote updates internal statistics for this peer by recording the vote.
+// It returns the total number of votes (1 per block). This essentially means
+// the number of blocks for which peer has been sending us block parts.
```
Comment (ebuchman) [inline consensus/reactor.go]: `sending us block parts` -> `sending us votes`

File: consensus/reactor.go [hunk 3]
```diff
@@ -1055,6 +1111,36 @@ func (ps *PeerState) ensureVoteBitArrays(height int64, numValidators int) {
...
+func (ps *PeerState) RecordVote(vote *types.Vote) int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	if ps.stats.lastVoteHeight == vote.Height {
+		return ps.stats.votes
+	}
+
```
Comment (ebuchman) [inline consensus/reactor.go]: dont we need to set lastVoteHeight now ?

File: consensus/reactor.go [hunk 4]
```diff
@@ -1055,6 +1111,36 @@ func (ps *PeerState) ensureVoteBitArrays(height int64, numValidators int) {
...
+// It returns the total number of block parts (1 per block). This essentially means
+// the number of blocks for which peer has been sending us block parts.
+func (ps *PeerState) RecordBlockPart(bp *BlockPartMessage) int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	if ps.stats.lastBlockPartHeight == bp.Height {
+		return ps.stats.blockParts
+	}
+
```
Comment (ebuchman) [inline consensus/reactor.go]: dont we need to set lastBlockPartHeight now ?

File: blockchain/reactor.go (hunk 1 above)
Comment (melekes) [reply blockchain/reactor.go]: i think so, yeah. it does not explain a) why it's default? b) why 1000 is a default? c) is it supposed to be used for all channels?

**Comment by ebuchman** [comment]
Let's rebase this so new CI runs ? 

Also which should be merged first this or #1282  ?

**Comment by melekes** [comment]
> Also which should be merged first this or #1282 ?

does not matter

**Comment by melekes** [comment]
@ebuchman rebased

File: consensus/reactor.go [hunk 5]
```diff
@@ -1055,6 +1073,37 @@ func (ps *PeerState) ensureVoteBitArrays(height int64, numValidators int) {
 	}
 }
 
+// RecordVote updates internal statistics for this peer by recording the vote.
+// It returns the total number of votes (1 per block). This essentially means
+// the number of blocks for which peer has been sending us votes.
+func (ps *PeerState) RecordVote(vote *types.Vote) int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
```
Comment (xla) [inline consensus/reactor.go]: do  we need the mutex here ?

**Comment by ebuchman** [comment]
Follow up in https://github.com/tendermint/tendermint/issues/1317

**Comment by ebuchman** [comment]
Also this needs tests. Might be tough to test given the current setup of consensus testing so worst case we leave that as an open issue

**Pull request merged**